pkg/client/restclient/slightly_sticky_provider.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restclient
+
+import (
+	"math/rand"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util/flowcontrol"
+)
+
+func newSlightlyStickyProvider(hosts []*url.URL) *slightlyStickyProvider {
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+	return &slightlyStickyProvider{
+		hosts:       hosts,
+		cur:         rng.Intn(len(hosts)),
+		rng:         rng,
+		ratelimiter: flowcontrol.NewTokenBucketRateLimiter(10, 0),
+	}
+}
+
+type slightlyStickyProvider struct {
+	sync.RWMutex
+	hosts       []*url.URL
+	cur         int
+	rng         *rand.Rand
+	ratelimiter flowcontrol.RateLimiter
+}
+
+func (s *slightlyStickyProvider) get() *url.URL {
+	s.RLock()
+	defer s.RUnlock()
+	return s.hosts[s.cur]
+}
+
+func (s *slightlyStickyProvider) next() {
+	s.RLock()
+	defer s.RUnlock()
+	s.cur = s.rng.Intn(len(s.hosts))
+	s.ratelimiter = flowcontrol.NewTokenBucketRateLimiter(10, 0)
```

Comment:
The algorithm is: rate limit the connection errors to 10 per second. If that rate is exceeded move to the next apiserver and reset the token bucket. The juju ratelimiter has no method to reset internal state but this is equivalent.